include/v4l2_camera/rate_bound_status.hpp

```diff
@@ -0,0 +1,281 @@
+#define RATE_BOUND_STATUS_HPP_
+
+#include <diagnostic_updater/diagnostic_updater.hpp>
+#include <diagnostic_msgs/msg/diagnostic_status.hpp>
+
+#include <iomanip>
+#include <sstream>
+#include <variant>
+#include <optional>
+#include <mutex>
+#include <stdexcept>
+
+namespace rate_bound_status
+{
+/**
+ * \brief A structure that holds the constructor parameters for the
+ * RateBoundStatus class.
+ */
+struct RateBoundStatusParam
+{
+  RateBoundStatusParam(const double min_freq, const double max_freq)
+      : min_frequency(min_freq), max_frequency(max_freq){}
+
+  double min_frequency;
+  double max_frequency;
+};
+
+/**
+ * \brief Diagnostic task to monitor the interval between events.
+ *
+ * This diagnostic task monitors the difference between consecutive events,
+ * and creates corresponding diagnostics. This task categorize observed intervals into
+ * OK/WARN/ERROR according to the value ranges passed via constructor arguments
+ */
+class RateBoundStatus : public diagnostic_updater::DiagnosticTask
+{
+private:
+  // Helper struct to express state machine nodes
+  struct StateBase
+  {
+    StateBase(const unsigned char lv, const std::string m)
+        : level(lv), num_observation(1), msg(m) {}
+
+    unsigned char level;
+    size_t num_observation;
+    std::string msg;
+  };
+
+  struct Stale : public StateBase
+  {
+    Stale()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::STALE,
+                    "Topic has not been received yet") {}
+  };
+
+  struct Ok : public StateBase
+  {
+    Ok()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::OK,
+                    "Rate is reasonable") {}
+  };
+
+  struct Warn : public StateBase
+  {
+    Warn()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::WARN,
+                    "Rate is within warning range") {}
+  };
+
+  struct Error : public StateBase
+  {
+    Error()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::ERROR,
+                    "Rate is out of valid range") {}
+  };
+
+  using StateHolder = std::variant<Stale, Ok, Warn, Error>;
+
+ public:
+  /**
+   * \brief Constructs RateBoundstatus, which inherits diagnostic_updater::DiagnosticTask.
+   *
+   * \param ok_params The pair of min/max frequency for the topic rate to be recognized as "OK".
+   * \param warn_params The pair of min/max frequency for the topic rate to be recognized as "WARN".
+   * These values should have a wider range than `ok_params`.
+   * \param num_frame_transition The number of the successive observations for the status
+   * transition. E.g., the status will not be changed from OK to WARN until successive
+   * `num_frame_transition` WARNs are observed.
+   * \param immediate_error_report If true (default), errors related to the rate bounds will be
+   * reported immediately once it is observed; otherwise, hysteresis manner using
+   * `num_frame_transition` will be adopted
+   * \param name The arbitral string to be assigned for this diagnostic task.
```

Comment:
I think you mean arbitrary?
```suggestion
   * \param name The arbitrary string to be assigned for this diagnostic task.
```

Reply:
Thank you for pointing this out. Yes, as you guessed, I mean "arbitrary".